components/node-labeler/cmd/run.go

```diff
@@ -249,6 +262,150 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
+			wsOld := e.ObjectOld.(*workspacev1.Workspace)
+			ws := e.ObjectNew.(*workspacev1.Workspace)
+			if wsOld.Status.Runtime == nil && ws.Status.Runtime != nil {
+				return true
+			}
+
+			// if we've seen runtime info before, there's no need to reconcile again
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+	}
+}
+
+func (wc *WorkspaceCountController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log.WithField("request", req.NamespacedName.String()).Info("WorkspaceCountController reconciling")
+
+	var ws workspacev1.Workspace
+	if err := wc.Get(ctx, req.NamespacedName, &ws); err != nil {
+		if !errors.IsNotFound(err) {
+			log.WithError(err).WithField("workspace", req.NamespacedName).Error("unable to fetch Workspace")
+			return ctrl.Result{}, err
+		}
+		// If workspace not found, do a full reconciliation
+		log.WithField("workspace", req.NamespacedName).Info("Workspace not found, reconciling all nodes")
+		return wc.reconcileAllNodes(ctx)
+	}
+
+	if ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != "" {
+		var workspaceList workspacev1.WorkspaceList
+		if err := wc.List(ctx, &workspaceList); err != nil {
+			log.WithError(err).Error("failed to list workspaces")
+			return ctrl.Result{}, err
+		}
+
+		count := 0
+		nodeName := ws.Status.Runtime.NodeName
+		for _, ws := range workspaceList.Items {
+			if ws.Status.Runtime != nil &&
+				ws.Status.Runtime.NodeName == nodeName &&
+				ws.DeletionTimestamp.IsZero() {
+				count++
+			}
+		}
+
+		if err := wc.updateNodeAnnotation(ctx, nodeName, count); err != nil {
+			return ctrl.Result{}, err
+		}
+		log.WithField("node", nodeName).WithField("count", count).Info("updated node annotation")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (wc *WorkspaceCountController) reconcileAllNodes(ctx context.Context) (ctrl.Result, error) {
+	var workspaceList workspacev1.WorkspaceList
+	if err := wc.List(ctx, &workspaceList); err != nil {
+		log.WithError(err).Error("failed to list workspaces")
+		return ctrl.Result{}, err
+	}
+
+	workspaceCounts := make(map[string]int)
+	for _, ws := range workspaceList.Items {
+		if ws.Status.Runtime != nil &&
+			ws.Status.Runtime.NodeName != "" &&
+			ws.DeletionTimestamp.IsZero() {
+			workspaceCounts[ws.Status.Runtime.NodeName]++
+		}
+	}
+
+	var nodes corev1.NodeList
+	if err := wc.List(ctx, &nodes); err != nil {
+		log.WithError(err).Error("failed to list nodes")
+		return ctrl.Result{}, err
+	}
+
+	for _, node := range nodes.Items {
+		count := workspaceCounts[node.Name]
+		if err := wc.updateNodeAnnotation(ctx, node.Name, count); err != nil {
+			log.WithError(err).WithField("node", node.Name).Error("failed to update node")
+			continue
+		}
+		log.WithField("node", node.Name).WithField("count", count).Info("updated node annotation")
+	}
+
+	return ctrl.Result{}, nil
+}
```

Comment:
This is a placeholder. I am not reviewing this until the above question about relying on the SyncPeriod to reconcile all workspaces is handled.